src/hotspot/share/opto/predicates.cpp

```diff
@@ -1081,11 +1087,22 @@ void TargetLoopPredicateChain::rewire_to_target_chain_head(IfTrueNode* template_
 
 ClonePredicateToTargetLoop::ClonePredicateToTargetLoop(LoopNode* target_loop_head, const NodeInLoopBody& node_in_loop_body,
                                                        PhaseIdealLoop* phase)
-    : _old_target_loop_entry(target_loop_head->in(LoopNode::EntryControl)),
-      _target_loop_predicate_chain(target_loop_head, phase),
+    : _target_loop_head(target_loop_head),
+      _target_outer_loop_head(target_loop_head->skip_strip_mined()),
+      _old_target_loop_entry(_target_outer_loop_head->in(LoopNode::EntryControl)),
+      _target_loop_predicate_chain(_target_outer_loop_head, phase),
       _node_in_loop_body(node_in_loop_body),
       _phase(phase) {}
 
+// Clones the provided Template Assertion Predicate to the head of the current predicate chain at the target loop.
+void ClonePredicateToTargetLoop::clone_template_assertion_predicate(
+    const TemplateAssertionPredicate& template_assertion_predicate) {
+  TemplateAssertionPredicate cloned_template_assertion_predicate =
+      template_assertion_predicate.clone(_old_target_loop_entry, _target_loop_head->as_CountedLoop(), _phase);
```

Comment:
Moved method from `.hpp` file here because of incomplete `CountedLoopNode` when calling `as_CountedLoop()`.